### ISSUE TITLE
Updater crashes on finish

### ISSUE THREAD

**Issue by 305jeff** [Epix37/Hearthstone-Deck-Tracker#526]
See here: http://puu.sh/e6lr1/e1c573d236.png

This has happened with a few of the recent updates...not 0.7.1 though.

Thanks!

**Comment by Epix37** [comment]
Weird, not quite sure what that could be.

**Comment by askarmuk** [comment]
I have updated manually to 0.7.3 and now it crashes at the start:

########## 05.01.2015 22:54:24 ##########
System.Collections.Generic.KeyNotFoundException: Данный ключ отсутствует в словаре.
   в System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   в Hearthstone_Deck_Tracker.Hearthstone.Game.get_IsMulliganDone()
   в Hearthstone_Deck_Tracker.HsLogReader.TagChange(String rawTag, Int32 id, String rawValue, Boolean isRecursive)
   в Hearthstone_Deck_Tracker.HsLogReader.Analyze(String log)
   в Hearthstone_Deck_Tracker.HsLogReader.<ReadFileAsync>d__1.MoveNext()
--- Конец трассировка стека из предыдущего расположения, где возникло исключение ---
   в System.Runtime.CompilerServices.AsyncMethodBuilderCore.<ThrowAsync>b__4(Object state)
   в System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   в MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
[22:54:21] : Checking for updates...
[22:54:23] : Copied log.config to C:\Users\временная\AppData\Local\Blizzard\Hearthstone\log.config (AlwaysOverwriteLogConfig)
[22:54:23] Hearthstone: Done loading localized card database (ruRU)
[22:54:23] Hearthstone: Done loading card database (enUS)
[22:54:23] : >>>>>>>>>>> Reset <<<<<<<<<<<
[22:54:23] OpponentSecrets: Cleared secrets

########## 05.01.2015 22:54:34 ##########
System.Collections.Generic.KeyNotFoundException: Данный ключ отсутствует в словаре.
   в System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   в Hearthstone_Deck_Tracker.Hearthstone.Game.get_IsMulliganDone()
   в Hearthstone_Deck_Tracker.HsLogReader.TagChange(String rawTag, Int32 id, String rawValue, Boolean isRecursive)
   в Hearthstone_Deck_Tracker.HsLogReader.Analyze(String log)
   в Hearthstone_Deck_Tracker.HsLogReader.<ReadFileAsync>d__1.MoveNext()
--- Конец трассировка стека из предыдущего расположения, где возникло исключение ---
   в System.Runtime.CompilerServices.AsyncMethodBuilderCore.<ThrowAsync>b__4(Object state)
   в System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   в MS.Internal.Threading.ExceptionFilterHelper.TryCatchWhen(Object source, Delegate method, Object args, Int32 numArgs, Delegate catchHandler)
[22:54:33] : Checking for updates...
[22:54:34] : Copied log.config to C:\Users\временная\AppData\Local\Blizzard\Hearthstone\log.config (AlwaysOverwriteLogConfig)
[22:54:34] Hearthstone: Done loading localized card database (ruRU)
[22:54:34] Hearthstone: Done loading card database (enUS)
[22:54:34] : >>>>>>>>>>> Reset <<<<<<<<<<<
[22:54:34] OpponentSecrets: Cleared secrets

**Comment by Epix37** [comment]
Try closing the Tracker and Hearthstone, then go to [hearthstone dir]/Hearthstone_Data, clear the output_log.txt file, save and start everything again.

**Comment by askarmuk** [comment]
It works now.

**Comment by 305jeff** [comment]
This just happened again with the latest update. See the link for a screenshot:

http://puu.sh/fKGRD/5bcedd2805.png

**Comment by Epix37** [comment]
Possibly related to #668 (permission issues)? If so this should be fixed after the next update.

**Comment by dvide** [comment]
I think this might be because file handles could still be open after enumerating through the temp files to copy them. When Updater tries to recursively delete the temp folder, it fails because it can't delete a file with a handle still open, and then the directory isn't empty, etc, causing this issue. I've created a pull request to at least stop it from crashing when this happens. It won't matter anyway if the temp folder doesn't get cleaned up, so we can get away with just fixing the symptom. Either way, the folder should get deleted next time the Updater starts (by then the file handles should have been released).
